utilities/transactions/lock/range/range_tree/range_tree_lock_manager.cc

```diff
@@ -0,0 +1,572 @@
+                                     const Endpoint& end_endp, Env*,
+                                     bool exclusive) {
+  toku::lock_request request;
+  request.create(mutex_factory_);
+  DBT start_key_dbt, end_key_dbt;
+
+  TEST_SYNC_POINT("RangeTreeLockManager::TryRangeLock:enter");
+  std::string start_key;
+  std::string end_key;
+  serialize_endpoint(start_endp, &start_key);
+  serialize_endpoint(end_endp, &end_key);
+
+  toku_fill_dbt(&start_key_dbt, start_key.data(), start_key.size());
+  toku_fill_dbt(&end_key_dbt, end_key.data(), end_key.size());
+
+  // Use the txnid as "extra" in the lock_request. Then, KillLockWait()
+  // will be able to use kill_waiter(txn_id) to kill the wait if needed
+  // TODO: KillLockWait is gone and we are no longer using
+  // locktree::kill_waiter call. Do we need this anymore?
+  TransactionID wait_txn_id = txn->GetID();
+
+  auto lt = get_locktree_by_cfid(column_family_id);
+
+  request.set(lt, (TXNID)txn, &start_key_dbt, &end_key_dbt,
+              exclusive ? toku::lock_request::WRITE : toku::lock_request::READ,
+              false /* not a big txn */, (void*)wait_txn_id);
+
+  // This is for "periodically wake up and check if the wait is killed" feature
+  // which we are not using.
+  uint64_t killed_time_msec = 0;
+  uint64_t wait_time_msec = txn->GetLockTimeout();
+
+  // convert microseconds to milliseconds
+  if (wait_time_msec != (uint64_t)-1)
+    wait_time_msec = (wait_time_msec + 500) / 1000;
+
+  std::vector<RangeDeadlockInfo> di_path;
+  request.m_deadlock_cb = [&](TXNID txnid, bool is_exclusive,
+                              const DBT* start_dbt, const DBT* end_dbt) {
+    EndpointWithString start;
+    EndpointWithString end;
+    deserialize_endpoint(start_dbt, &start);
+    deserialize_endpoint(end_dbt, &end);
+
+    di_path.push_back({((PessimisticTransaction*)txnid)->GetID(),
+                       column_family_id, is_exclusive, std::move(start),
+                       std::move(end)});
+  };
+
+  request.start();
+
+  /*
+    If we are going to wait on the lock, we should set appropriate status in
+    the 'txn' object. This is done by the SetWaitingTxn() call below.
+    The API we are using are MariaDB's wait notification API, so the way this
+    is done is a bit convoluted.
+    In MyRocks, the wait details are visible in I_S.rocksdb_trx.
+  */
+  std::string key_str(start_key.data(), start_key.size());
+  struct st_wait_info {
+    PessimisticTransaction* txn;
+    uint32_t column_family_id;
+    std::string* key_ptr;
+    autovector<TransactionID> wait_ids;
+    bool done = false;
+
+    static void lock_wait_callback(void* cdata, TXNID /*waiter*/,
+                                   TXNID waitee) {
+      TEST_SYNC_POINT("RangeTreeLockManager::TryRangeLock:WaitingTxn");
+      auto self = (struct st_wait_info*)cdata;
+      // we know that the waiter is self->txn->GetID() (TODO: assert?)
+      if (!self->done) {
```

Comment:
will the callback be called multiple times? Maybe each time on a new waitee?

Reply:
What if txn 0 needs to wait for txn 1, then when txn 1 commits, txn 0 needs to wait for txn 2?

Reply:
Ok, this uncovered a fairly big issue.  locktree library had an API for reporting lock waits in form  "transaction X is now waiting on transaction Y", and RangeTreeLockManager was trying to use it.  But, it didn't provide all the required information, and also RangeTreeLockManager was not using it correctly.  Now all fixed, and test coverage is added.